python/ray/autoscaler/_private/vsphere/node_provider.py

```diff
@@ -0,0 +1,726 @@
+
+    def create_node_tag(self, ray_node_tag, category_id):
+        cli_logger.info(f"Creating {ray_node_tag} tag")
+        tag_spec = self.vsphere_sdk_client.tagging.Tag.CreateSpec(
+            ray_node_tag, "Ray node tag", category_id
+        )
+        tag_id = None
+        try:
+            tag_id = self.vsphere_sdk_client.tagging.Tag.create(tag_spec)
+        except ErrorClients.Unauthorized as e:
+            cli_logger.abort(f"Unathorised to create the tag. Exception: {e}")
+        except Exception as e:
+            cli_logger.abort(e)
+
+        cli_logger.info(f"Tag {tag_id} created")
+        return tag_id
+
+    def get_category(self):
+        for id in self.vsphere_sdk_client.tagging.Category.list():
+            if (
+                self.vsphere_sdk_client.tagging.Category.get(id).name
+                == Constants.NODE_CATEGORY
+            ):
+                return id
+        return None
+
+    def create_category(self):
+        # Create RAY_NODE category. This category is associated with VMs and supports
+        # multiple tags e.g. "Ray-Head-Node, Ray-Worker-Node-1 etc."
+        cli_logger.info(f"Creating {Constants.NODE_CATEGORY} category")
+        category_spec = self.vsphere_sdk_client.tagging.Category.CreateSpec(
+            name=Constants.NODE_CATEGORY,
+            description="Identifies Ray head node and worker nodes",
+            cardinality=CategoryModel.Cardinality.MULTIPLE,
+            associable_types=set(),
+        )
+        category_id = None
+
+        try:
+            category_id = self.vsphere_sdk_client.tagging.Category.create(category_spec)
+        except ErrorClients.Unauthorized as e:
+            cli_logger.abort(f"Unathorised to create the category. Exception: {e}")
+        except Exception as e:
+            cli_logger.abort(e)
+
+        cli_logger.info(f"Category {category_id} created")
+
+        return category_id
+
+    def terminate_node(self, node_id):
+        if node_id is None:
+            return
+
+        status = self.vsphere_sdk_client.vcenter.vm.Power.get(node_id)
+
+        if status.state != HardPower.State.POWERED_OFF:
+            self.vsphere_sdk_client.vcenter.vm.Power.stop(node_id)
+            cli_logger.info("vm.Power.stop({})".format(node_id))
+
+        self.vsphere_sdk_client.vcenter.VM.delete(node_id)
+        cli_logger.info("Deleted vm {}".format(node_id))
+
+        # Pop node_id from cached_nodes and tag_cache only if not present
+        if node_id in self.cached_nodes:
+            self.cached_nodes.pop(node_id)
+
+        if node_id in self.tag_cache:
+            self.tag_cache.pop(node_id)
+
+    def terminate_nodes(self, node_ids):
+        if not node_ids:
+            return
+
+        for node_id in node_ids:
+            self.terminate_node(node_id)
+
+    def _get_node(self, node_id):
+        """Refresh and get info for this node, updating the cache."""
+        self.non_terminated_nodes({})  # Side effect: updates cache
```

Comment:
Good call-out, does it make sense to have an helper function that explicitly updates the cache, instead of relying on this side-effect? As written it looks a little weird.

Reply:
We don't need to refresh the cache here because the cache contains all the non-terminated node. But here _get_node should be able to return terminated nodes.

Reply:
I think if we add a refresh_cache function, most of the logic will be same with current "non_terminated_nodes". It may be not good to add duplicate logic.

As we have no where in our code need to refresh the cache for now, we didn't add such a function but we will do if we have such requirement in the future.